R/getPrevalence.R

```diff
@@ -300,17 +346,81 @@ setMethod("getRareTaxa", signature = c(x = "SummarizedExperiment"),
     function(x, rank = NULL, ...){
         # Gets the prevalent taxa
         prev_taxa <- getPrevalentTaxa(x, rank = rank, ...)
+        # If agglomeration was done
         if( !is.null(rank) ){
             # Gets names from specified taxonomic level
             taxa <- rowData(x)[[rank]]
-        } else{
-            # Gets rownames if agglomeration is not done
-            taxa <- rownames(x)
+        } else {
+            # If rownames is not NULL
+            if( !is.null(rownames(x)) ){
+                # Gets rownames 
+                taxa <- rownames(x)
+            } else {
+                # Gets indices of taxa
+                taxa <- 1:nrow(tse)
+            }
         }
         unique(taxa[!is.na(taxa) & !(taxa %in% prev_taxa)])
     }
 )
 
+############################# subsetByPrevalentTaxa ############################
+
+#' @rdname getPrevalence
+#' @export
+setGeneric("subsetByPrevalentTaxa", signature = "x",
+           function(x, ...)
+               standardGeneric("subsetByPrevalentTaxa"))
+
+#' @rdname getPrevalence
+#' @export
+setMethod("subsetByPrevalentTaxa", signature = c(x = "SummarizedExperiment"),
+    function(x, rank = NULL, ...){
+        # If rank is not NULL
+        if( !is.null(rank) ){
```

Comment:
The check for missing rownames is not present, yet. Again, I would try putting it just before `getPrevalentTaxa`

Reply:
Does it need check for` rownames(x)`? This calls `getPrevalentTaxa`. `getPrevalentTaxa` returns taxa names or indices depending whether `rownames(x)` is NULL or not. Because of that subsetting should work

Reply:
That might be the case, but is not really a good structure. One of the mantras "fail early, fail often" can also be extended to "check early, check often"

In addition, I am not so sure, whether it really works. If `x` in `subsetByPrevalentTaxa` does not have rownames, the subsetting works "only" maybe. Even if you go in with "indices", as you have called them, they are actually a character vector and not a numeric vector (rownames are always `character`). Therefore the subsetting may only work, because the subsetting function maybe tries to convert the character vector to an integer one, but I am not so sure about it.

If you check the rownames in `subsetByPrevalentTaxa`, the rownames are set and the returned indices are actually the rownames of the object your are trying to subset. All in all, this is a much more robust solution.

And since you have to check for rownames in four functions, write an internal function for this check, so that you only have to write it once.